apis/Google.Cloud.AspNetCore.DataProtection.SecretManager/Google.Cloud.AspNetCore.DataProtection.SecretManager/CloudSecretManagerXmlRepository.cs

```diff
@@ -0,0 +1,120 @@
+using System.Linq;
+using System.Text;
+using System.Xml.Linq;
+using Secret = Google.Cloud.SecretManager.V1.Secret;
+
+namespace Google.Cloud.AspNetCore.DataProtection.SecretManager;
+
+/// <summary>
+/// Implementation of <see cref="IXmlRepository"/> that stores the protected elements in a Google Cloud Secret Manager.
+/// This class is configured by <see cref="GoogleCloudDataProtectionBuilderExtensions.PersistKeysToGoogleCloudSecretManager(Microsoft.AspNetCore.DataProtection.IDataProtectionBuilder, string, string)"/>
+/// (and other overloads).
+/// </summary>
+internal class CloudSecretManagerXmlRepository : IXmlRepository
+{
+    private readonly SecretManagerServiceClient _secretManagerServiceClient;
+    private readonly SecretName _secretName;
+    private readonly ProjectName _projectName;
+    private readonly SecretVersionName _version;
+
+    internal CloudSecretManagerXmlRepository(SecretManagerServiceClient client, string secretName, string projectId)
+    {
+        _secretManagerServiceClient = GaxPreconditions.CheckNotNull(client, nameof(client));
+        GaxPreconditions.CheckNotNull(secretName, nameof(secretName));
+        GaxPreconditions.CheckNotNull(projectId, nameof(projectId));
+        _projectName = new ProjectName(projectId);
+        _secretName = new SecretName(projectId, secretName);
+
+        // Version is populated in such a way that it always refers to the most latest version.
+        // For more information on version naming refer to
+        // https://cloud.google.com/dotnet/docs/reference/Google.Cloud.SecretManager.V1/latest/Google.Cloud.SecretManager.V1.AccessSecretVersionRequest#Google_Cloud_SecretManager_V1_AccessSecretVersionRequest_Name
+        _version = SecretVersionName.FromProjectSecretSecretVersion(projectId, secretName, "latest");
+    }
+
+    /// <inheritdoc />
+    public IReadOnlyCollection<XElement> GetAllElements()
+    {
+        if (!SecretExists())
+        {
+            return Array.Empty<XElement>();
+        }
+        var accessSecretVersionResponse = _secretManagerServiceClient.AccessSecretVersion(_version);
+
+        if (accessSecretVersionResponse != null)
+        {
+            var secretPayload = accessSecretVersionResponse.Payload.Data.ToStringUtf8();
+            XDocument document = XDocument.Parse(secretPayload);
+            var xElements = document.Root.Elements().ToList();
+            return xElements;
+        }
+        return Array.Empty<XElement>();
+    }
+
+    /// <inheritdoc />
+    public void StoreElement(XElement element, string friendlyName)
+    {
+        if (!SecretExists())
+        {
+            CreateSecret();
+        }
+        XDocument document = new XDocument(new XElement("root"));
+        document.Root.Add(element);
+        SecretPayload payload = new SecretPayload
+        {
+            Data = ByteString.CopyFrom(document.ToString(), Encoding.UTF8)
+        };
+        _secretManagerServiceClient.AddSecretVersion(_secretName, payload);
```

Comment:
Doesn't this overwrite anything that was previously there? I'd expect it to just add a new top-level element.

Compare this with https://github.com/dotnet/aspnetcore/blob/main/src/DataProtection/DataProtection/src/Repositories/FileSystemXmlRepository.cs - where multiple elements are basically stored as separate files.

I'd expect an integration test of:

- Create new secret
- Add element 1
- Add element 2
- Fetch all elements (both should be there)

Reply:
Got it Jon Thanks for pointing this out. I missed this part.
Will rectifie it.